### PR TITLE
fix(tracing): use psycopg-version-specific DSN parsers

### PR DESCRIPTION
**Pull request by yolken-rippling**
This change updates the DSN parsing logic in `ddtrace/ext/sql.py` to use separate functions for psycopg2 and psycopg3. This should help prevent issues when users are using both versions of psycopg simultaneously.

See https://github.com/DataDog/dd-trace-py/issues/9414 for more details.

## Checklist

- [X] Change(s) are motivated and described in the PR description
- [X] Testing strategy is described if automated tests are not included in the PR
- [X] Risks are described (performance impact, potential for breakage, maintainability)
- [X] Change is maintainable (easy to change, telemetry, documentation)
- [X] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [X] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [X] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [X] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [ ] Title is accurate
- [ ] All changes are related to the pull request's stated goal
- [ ] Description motivates each change
- [ ] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [ ] Testing strategy adequately addresses listed risks
- [ ] Change is maintainable (easy to change, telemetry, documentation)
- [ ] Release note makes sense to a user of the library
- [ ] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [ ] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/ext/sql.py [hunk 1]
```diff
@@ -44,32 +44,37 @@ def _dd_parse_pg_dsn(dsn):
 # Do not import from psycopg directly! This reference will be updated at runtime to use
 # a better implementation that is provided by the psycopg library.
 # This is done to avoid circular imports.
-parse_pg_dsn = _dd_parse_pg_dsn
+
+# Parser for psycopg2
+parse_pg_dsn2 = _dd_parse_pg_dsn
```
Comment (P403n1x87) [inline ddtrace/ext/sql.py:49]: Perhaps we can have a single global `_parse_pg_func` that we set via the import hook below, and then export a single helper, e.g. `parse_pg_dsn` that encapsulates the common logic
```python
        def parse_pg_dsn(conn):
            global _parse_pg_func

            if hasattr(conn, "dsn"):
                return _parse_pg_func(conn.dsn)
            elif hasattr(conn, "info") and hasattr(conn.info, "dsn"):
                return _parse_pg_func(conn.info.dsn)
            return None
```
?

Comment (yolken-rippling) [reply ddtrace/ext/sql.py:49]: Sure, although I think we still need separate `_parse_pg_func`s, one for each psycopg version? In any case, changed it to pull the logic around getting the DSN from the connection into a shared utility function. Thanks!

File: ddtrace/contrib/psycopg/connection.py [hunk 2]
```diff
@@ -63,21 +63,22 @@ def patch_conn(conn, traced_conn_cls, pin=None):
             _patch_extensions(extensions_to_patch)
 
     c = traced_conn_cls(conn)
-
-    # if the connection has an info attr, we are using psycopg3
-    if hasattr(conn, "dsn"):
-        dsn = sql.parse_pg_dsn(conn.dsn)
+    dsn = sql.parse_pg_conn_dsn(conn)
+
+    if dsn:
```
Comment (P403n1x87) [inline ddtrace/contrib/psycopg/connection.py:68]:
```suggestion
    if dsn is not None:
```

**Review by P403n1x87 — COMMENTED** [review]
Just a nit, otherwise LGTM. Thanks for the contribution!

File: ddtrace/contrib/psycopg/connection.py (hunk 2 above)
Comment (yolken-rippling) [reply ddtrace/contrib/psycopg/connection.py]: Fixed!

**Review by P403n1x87 — APPROVED** [review]

**Review by emmettbutler — APPROVED** [review]

**Comment by P403n1x87** [comment]
Because this is a fix for an issue, we need a [release note](https://ddtrace.readthedocs.io/en/stable/releasenotes.html?highlight=reno#generating-a-release-note) to go with the change.

**Comment by yolken-rippling** [comment]
> Because this is a fix for an issue, we need a [release note](https://ddtrace.readthedocs.io/en/stable/releasenotes.html?highlight=reno#generating-a-release-note) to go with the change.

Done.

**Pull request closed without merging**